File: ios/SuiteBDE/UI/Components/AlertCase.swift
//
//  AlertCase.swift
//  BDE
//
//  Created by Nathan Fallet on 18/05/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import shared

extension AlertCase: Identifiable {
    
    public var id: String { name }
    
}

func constructAlertCase(
    discardEdit: @escaping () -> Void,
    deleteAccount: @escaping () -> Void = {}
) -> ((AlertCase) -> Alert) {
    return { item in
        switch item {
        case .cancelling:
            return Alert(
                title: Text("Les modifications n'ont pas été enregistrées"),
                message: Text("Etes vous sûr de vouloir quitter ?"),
                primaryButton: .cancel(Text("Quitter sans enregistrer"), action: discardEdit),
                secondaryButton: .default(Text("Revenir sur l'édition"))
            )
        case .deleting:
            return Alert(
                title: Text("Es tu sûr de vouloir supprimer ton compte ?"),
                message: Text("Cette action est irréversible et tu perdras toutes tes données."),
                primaryButton: .cancel(Text("Annuler")),
                secondaryButton: .destructive(Text("Supprimer mon compte"), action: deleteAccount)
            )
        case .saved:
            return Alert(
                title: Text("Les modifications ont bien été enregistrées")
            )
        default:
            return Alert(
                title: Text("Error")
            )
        }
    }
}

File: android/src/main/java/me/nathanfallet/suitebde/ui/components/AlertCase.kt
package me.nathanfallet.suitebde.ui.components

import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import me.nathanfallet.suitebde.models.application.AlertCase

@Composable
fun AlertCaseDialog(
    alertCase: AlertCase?,
    onDismissRequest: () -> Unit,
    discardEdit: () -> Unit,
    deleteAccount: () -> Unit = {},
) {
    when (alertCase) {
        AlertCase.CANCELLING -> {
            AlertDialog(
                onDismissRequest = onDismissRequest,
                title = { Text("Les modifications n'ont pas été enregistrées") },
                text = { Text("Etes vous sûr de vouloir quitter ?") },
                confirmButton = {
                    Button(onClick = {
                        discardEdit()
                        onDismissRequest()
                    }) {
                        Text("Quitter sans enregistrer")
                    }
                },
                dismissButton = {
                    Button(onClick = onDismissRequest) {
                        Text("Revenir sur l'édition")
                    }
                }
            )
        }

        AlertCase.DELETING -> {
            AlertDialog(
                onDismissRequest = onDismissRequest,
                title = { Text("Es tu sûr de vouloir supprimer ton compte ?") },
                text = { Text("Cette action est irréversible et tu perdras toutes tes données.") },
                confirmButton = {
                    Button(onClick = {
                        deleteAccount()
                        onDismissRequest()
                    }) {
                        Text("Supprimer mon compte")
                    }
                },
                dismissButton = {
                    Button(onClick = onDismissRequest) {
                        Text("Annuler")
                    }
                }
            )
        }

        AlertCase.SAVED -> {
            AlertDialog(
                onDismissRequest = onDismissRequest,
                title = { Text("Les modifications ont bien été enregistrées") },
                confirmButton = {
                    Button(onClick = onDismissRequest) {
                        Text("OK")
                    }
                }
            )
        }

        // TODO: Error

        else -> {} // Nothing in case it's null
    }
}
